### PR TITLE
feat(Pipelines) synthetic dataframe pipeline

### PR DESCRIPTION
**Pull request by ArslanSaleem**
## Summary by CodeRabbit

New Features:
- Introduced a new pipeline system in the `pandasai` library for generating synthetic dataframes.
- Added new classes `AbstractPipeline`, `Pipeline`, `PipelineContext`, `BaseLogicUnit`, `BaseCodeExecutor`, `PromptExecution`, `GenerateSDFPipeline`, `SDFCodeExecutor`, and `SyntheticDataframePrompt` to support the pipeline system.
- Added a new exception class `UnSupportedLogicUnit` to handle unsupported logic units in the pipeline.

Refactor:
- Updated import statements and class inheritance in various files for better code organization and readability.

Tests:
- Added a new test suite for the `Pipeline` class to ensure its proper functioning.

Chores:
- Updated the `examples/using_pipeline.py` file to demonstrate the usage of the new pipeline system.

### CODE REVIEW

File: pandasai/pipelines/pipeline_context.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    _dfs: List[Union[DataFrameType, SmartDataframe]]
+    _memory: Memory
+    _skills: SkillsManager
+
+    def __init__(
+        self,
+        dfs: List[Union[DataFrameType, SmartDataframe]],
+        memory: Memory = None,
+        skills: SkillsManager = None,
```
Comment (gventuri) [inline pandasai/pipelines/pipeline_context.py:22]: Maybe we can put the `cache` as well as part of the context?

File: pandasai/pipelines/pipeline_context.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+class PipelineContext:
+    """
+    Pass Context to the pipeline which is accessible to each step via kwargs
+    """
+
+    _dfs: List[Union[DataFrameType, SmartDataframe]]
+    _memory: Memory
+    _skills: SkillsManager
+
```
Comment (gventuri) [inline pandasai/pipelines/pipeline_context.py:19]:
```suggestion
_cache: Cache
```

**Comment by gventuri** [comment]
@ArslanSaleem could you handle the conflicts, then we are ready to merge!

**Comment by ArslanSaleem** [comment]
Opening another PR for this change

**Pull request closed without merging**